qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+    if weights is None:
+        wsigma = sigma
+    else:
+        wsigma = np.zeros(ydata.size)
+        if sigma is None:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = 1 / np.sqrt(weights[i])
+        else:
+            for i in range(num_funcs):
+                wsigma[idxs[i]] = sigma[idxs[i]] / np.sqrt(weights[i])
+
+    # Define multi-objective function
+    def f(x, *params):
+        y = np.zeros(x.size)
+        for i in range(num_funcs):
+            xi = x[idxs[i]]
+            yi = funcs[i](xi, *params)
+            y[idxs[i]] = yi
+        return y
+
+    # Run linearized curve_fit
+    analysis_result = curve_fit(f, xdata1d, ydata, p0, sigma=wsigma, **kwargs)
+
+    return analysis_result
+
+
+def curve_fit_data(
+    data: List[Dict[str, any]], data_processor: Callable, x_key: str = "xval", **filters
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Return array of (x, y, sigma) data for curve fitting.
+
+    Args
+        data: list of circuit data dictionaries containing counts.
+        data_processor: callable for processing data to y, sigma
+        x_key: key for extracting xdata value from metadata (Default: "xval").
+        filters: additional kwargs to filter metadata on.
+
+    Returns:
+        tuple: ``(x, y, sigma)`` tuple of arrays of x-values,
+               y-values, and standard deviations of y-values.
+
+    Raises:
+        QiskitError: if input data is not level-2 measurement.
+    """
+    filtered_data = filter_data(data, **filters)
+    size = len(filtered_data)
+    xdata = np.zeros(size, dtype=int)
+    ydata = np.zeros(size, dtype=float)
+    ydata_var = np.zeros(size, dtype=float)
+
+    for i, datum in enumerate(filter_data):
+        metadata = datum["metadata"]
+        xdata[i] = metadata[x_key]
+        y_mean, y_var = data_processor(datum)
+        ydata[i] = y_mean
+        ydata_var[i] = y_var
+
+    return xdata, ydata, np.sqrt(ydata_var)
+
+
+def multi_curve_fit_data(
+    data: List[Dict[str, any]],
+    data_processor: Callable,
+    x_key: str = "xval",
+    series_key: str = "series",
+    **filters,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """Return array of (x, y, sigma) data for curve fitting.
+
+    Args
+        data: list of circuit data dictionaries.
+        data_processor: callable for processing data to y, sigma
+        x_key: key for extracting xdata value from metadata (Default: "xval").
+        series_key: key for extracting series value from metadata (Default: "series").
+        filters: additional kwargs to filter metadata on.
+
+    Returns:
+        tuple: ``(x, y, sigma)`` tuple of arrays of x-values,
+               y-values, and standard deviations of y-values.
+
+    Raises:
+        QiskitError: if input data is not level-2 measurement.
+    """
+    filtered_data = filter_data(data, **filters)
+    size = len(filtered_data)
+    xdata = np.zeros((size, 2), dtype=float)
```

Comment:
Why we cannot return index separately? This enforces series to be float value, but actually we may want to use string representation. For example, Hamiltonian tomography, this would be `meas_basis \in x, y, z`. We can map them to `[0, 1, 2]` but series should be directly used as a label when we generate a plot. Currently user need to manage mapping between labels and lines.

Reply:
Do you mean return two separate 1D arrays for x-data (float), and series-data (int)? That should be possible if I change the signature of `multi_curve_fit` to pass series in as a separate kwarg to xdata.

I think the value needs to be an integer though because it is used as the index of the component function in multi curve fit. A string value is ambiguous for ordering so you would need to assign an integer value for each string to assign it to a series which could be done at a higher level than these functions.

Reply:
okey, that make sense. I prefer having index as separate `dtype=int` array to make it clear. Also having that description in docstring would be great.